### ISSUE TITLE
Creating a realm instance fails with a cannot allocate error even when there is more than enough memory

### ISSUE THREAD

**Issue by monolithic-adam** [realm/realm-cocoa#6104]
## Goals
Eliminate realm configuration errors

## Expected Results
No Crashes

## Actual Results
Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=io.realm Code=1 "fcntl() inside prealloc() failed allocating 1184 bytes, new_size=1184, cur_size=0, allocated_size=0, event: Operation not permitted" UserInfo={NSLocalizedDescription=fcntl() inside prealloc() failed allocating 1184 bytes, new_size=1184, cur_size=0, allocated_size=0, event: Operation not permitted, Error Code=1}: file /BuildRoot/Library/Caches/com.apple.xbs/Sources/swiftlang_Fall2018/swiftlang_Fall2018-1000.11.42/src/swift/stdlib/public/core/ErrorType.swift, line 184

## Steps to Reproduce
We cannot reproduce it locally but the issues are showing up in Crashylitics

## Code Sample
```swift
    public var realm: Realm {
        let realm: Realm
        do {
            realm = try Realm(configuration: customConfiguration)
        } catch let error as NSError {
            // Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=io.realm Code=1 "fcntl() inside prealloc() failed allocating 1184 bytes, new_size=1184, cur_size=0, allocated_size=0, event: Operation not permitted"
            // If the realm file cannot be opened, remove it
            RealmClient.removeRealmFile()
            realm = try! Realm(configuration: customConfiguration)
        }
        return realm
    }
```

## Version of Realm and Tooling
Realm framework version: 3.7.6

Realm Object Server version: ?

Xcode version: 10.1

iOS/OSX version: iOS 11 & 12

Dependency manager + version: Cocoapods 1.6.0

**Comment by tomsemi** [comment]
how to solve this problem?

**Comment by tomsemi** [comment]
@monolithic-adam do you know how to fix this bug?

**Comment by monolithic-adam** [comment]
> @monolithic-adam do you know how to fix this bug?

@tomsemi Sorry I don't... me and my team have been trying to figure out what causes it for awhile but we can't nail down the cause.

**Comment by tomsemi** [comment]
> > @monolithic-adam do you know how to fix this bug?
> 
> @tomsemi Sorry I don't... me and my team have been trying to figure out what causes it for awhile but we can't nail down the cause.

Realm framework version: 3.7.6
only crash happens in this version? the previous versions is OK?

**Comment by tomsemi** [comment]
> > @monolithic-adam do you know how to fix this bug?
> 
> @tomsemi Sorry I don't... me and my team have been trying to figure out what causes it for awhile but we can't nail down the cause.

we didn't get so many crash reports while we use the 2.8.0 version. how about your team?

**Comment by monolithic-adam** [comment]
> > > @monolithic-adam do you know how to fix this bug?
> > 
> > 
> > @tomsemi Sorry I don't... me and my team have been trying to figure out what causes it for awhile but we can't nail down the cause.
> 
> Realm framework version: 3.7.6
> only crash happens in this version? the previous versions is OK?
We have been getting it for a long time, it has been like that since around 6 months ago
It started after we switched from version 3.1.0 -> 3.7.6

**Comment by tgoyne** [comment]
This isn't an out-of-memory error. It's reporting that we got the error EPERM (Operation not permitted) when attempting to resize a file from 0 bytes to 1184 bytes.

What this actually means is unclear. The `fcntl()` man page doesn't list EPERM as a possible result for the call we're making, and it seems very unlikely to be due to file protection (the user would have to lock the device in between us creating the file and resizing it?).

The relevant code did change between 3.1.0 and 3.7.6. We added a call to `fcntl(F_PREALLOCATE)` prior to `ftruncate()` when resizing files as we found that was required to actually get an immediate error when the system had insufficient disk space to satisfy the request (previously we'd just get SIGBUS when trying to use the space). We've seen some APFS bugs with F_PREALLOCATE before, but not any that result in it failing with EPERM. Perhaps we just need to fall back to the non-preallocate codepath for that like we do with EINVAL?

**Comment by monolithic-adam** [comment]
> This isn't an out-of-memory error. It's reporting that we got the error EPERM (Operation not permitted) when attempting to resize a file from 0 bytes to 1184 bytes.
> 
> What this actually means is unclear. The `fcntl()` man page doesn't list EPERM as a possible result for the call we're making, and it seems very unlikely to be due to file protection (the user would have to lock the device in between us creating the file and resizing it?).
> 
> The relevant code did change between 3.1.0 and 3.7.6. We added a call to `fcntl(F_PREALLOCATE)` prior to `ftruncate()` when resizing files as we found that was required to actually get an immediate error when the system had insufficient disk space to satisfy the request (previously we'd just get SIGBUS when trying to use the space). We've seen some APFS bugs with F_PREALLOCATE before, but not any that result in it failing with EPERM. Perhaps we just need to fall back to the non-preallocate codepath for that like we do with EINVAL?

Is there something we should try? We are getting this where we first get a realm instance in didFinishLaunching if that helps.

**Comment by zzzworm** [comment]
We also meet this problem, and now it is our top crash reason.please fix it.

**Comment by purepure** [comment]
Our team also get in to this issue after upgrade Realm. Now we get 2k crash per day. It is crazy. The previous is more stable (with very few crash).
We follow some solutions but seem not find the solution. Our team cannot reproduce it locally so it is more difficult to fix.
Since the db is larger day by day, I am thinking to not using Realm locally.

**Comment by rogerluan** [comment]
We're facing this too. Was any work done towards this issue since this message https://github.com/realm/realm-cocoa/issues/6104#issuecomment-473974782, @tgoyne ?

**Comment by daymont87** [comment]
In our app it's happening every time when database getting size 2Gb++. I don't understand why developers can't just reproduce this crash and find crash reason

**Comment by ppamorim** [comment]
@daymont87 2Gb is too much for a database, please divide it into multiple Realm files and try to compress it as much as possible. Also please remove any file if you are including into it.

**Comment by hhongak** [comment]
so, it can't be fix?  @bmunkholm

**Comment by rogerluan** [comment]
Yeah I don't know what's the status of this 😕 Is this a nonrecoverable error? How can we avoid it?

**Comment by sainttail** [comment]
I also have this problem with realm 4.4.0 and it is the top of the crash in crashlytic
